src/re-parse.y

```diff
@@ -32,14 +33,17 @@ void yyerror(const char msg[]);
 %type <int_val> TOK_CHAR TOK_NUMBER
 %type <cce_val> TOK_CCE
 %type <ccl_val> TOK_CCL ccl full_ccl
-%type <mach_val> re singleton series string
+%type <mach_val> re singleton series string disjunction
 
 %destructor { delete $$; } <mach_val>
 
 %%
 flexrule	:	re
 			{ $1->AddAccept(1); zeek::detail::nfa = $1; }
 
+		|  disjunction
+			{ zeek::detail::nfa = $1; }
+
```

Comment:
Hey @vpax - I was staring at this for probably way too long, but am wondering: Can we instead re-use the `Specific_RE_Matcher::CompileSet()` that already exists and remove the `||` parsing "trick" in REs?

Mostly because I was right away trying whether `/||a||b/` is now a supported pattern. It's not in script land, but it is accepted for signature files and also accessible via the `RE_Matcher` API.

The regex `||a||b` is valid in other languages (tried Python, JavaScript, Ruby) and while it might not be super sensible, it does also match the empty string.

Was there a compelling reason not to use CompileSet()? I've worked out the patch below and it does pass your new test-case (not including re-parse.* revert for brevity). Unfortunately, I'm not well versed here, so I might be missing the obvious.

```--- a/src/RE.cc
+++ b/src/RE.cc
@@ -194,6 +194,10 @@ bool Specific_RE_Matcher::CompileSet(const string_list& set, const int_list& idx
        dfa = new DFA_Machine(nfa, EC());
        ecs = EC()->EquivClasses();
 
+       // dfa took ownership
+       Unref(nfa);
+       nfa = nullptr;
+
        return true;
        }
 
@@ -523,13 +527,17 @@ RE_DisjunctiveMatcher::RE_DisjunctiveMatcher(const std::vector<const RE_Matcher*
        {
        matcher = std::make_unique<detail::Specific_RE_Matcher>(detail::MATCH_EXACTLY);
 
-       std::string disjunction;
-       for ( auto re : REs )
-               disjunction += std::string("||") + re->PatternText();
+       zeek::detail::string_list sl;
+       zeek::detail::int_list il;
+
+       for ( const auto* re : REs )
+               {
+               sl.push_back(const_cast<char*>(re->PatternText()));
+               il.push_back(sl.size());
+               }
 
-       matcher->SetPat(disjunction.c_str());
-       auto status = matcher->Compile();
-       ASSERT(status);
+       if ( ! matcher->CompileSet(sl, il) )
+               reporter->FatalError("failed compile set for disjunctive matcher");
        }


```

Reply:
The key requirement is being able to associate different accepting numbers with the different elements of the disjunction. The place I saw to do that was in the RE parser. If `CompileSet` has the same effect, then yeah fine to go that route instead!

Reply:
Yeah, it's using the values provided in the idx list as accepting numbers for the individual REs (`il` in the patch above).

https://github.com/zeek/zeek/blob/4d3e9de3d309c77f0ee7a55447ad66eb6e6887e4/src/RE.cc#L180-L183